src/hooks.ts

```diff
@@ -0,0 +1,117 @@
+/**
+ * Copyright 2019-2020, Optimizely
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { useCallback, useContext, useEffect, useState } from 'react';
+import * as optimizely from '@optimizely/optimizely-sdk';
+import { getLogger } from '@optimizely/js-sdk-logging';
+
+import { setupAutoUpdateListeners } from './autoUpdate';
+import { VariableValuesObject, OnReadyResult } from './client'
+import { OptimizelyContext } from './Context';
+
+type UseFeatureState = {
+  isEnabled: Boolean,
+  variables: VariableValuesObject,
+};
+
+type ClientReady = Boolean;
+
+type UseFeatureOptions = {
+  autoUpdate?: Boolean,
+  timeout?: number
+};
+
+type UseFeatureOverrides = {
+  overrideUserId?: string,
+  overrideAttributes?: optimizely.UserAttributes,
+}
+
+interface UseFeature {
+  (
+    featureKey: string,
+    options?: UseFeatureOptions,
+    overrides?: UseFeatureOverrides,
+  ): [
+    UseFeatureState["isEnabled"],
+    UseFeatureState["variables"],
+    ClientReady
+  ]
+}
+
+/**
+ * A React Hook that retrieves the status of a feature flag and its variables, optionally
+ * auto updating those values based on underlying user or datafile changes.
+ */
+export const useFeature : UseFeature = (featureKey, options = {}, overrides = {}) => {
+  const { isServerSide, optimizely, timeout } = useContext(OptimizelyContext);
+  const finalReadyTimeout: number | undefined =
+      options.timeout !== undefined ? options.timeout : timeout;
+
+  // Helper function to return the current values for isEnabled and variables.
+  const getCurrentValues = useCallback(() => ({
+    isEnabled: optimizely ? optimizely.isFeatureEnabled(featureKey, overrides.overrideUserId, overrides.overrideAttributes) : false,
```

Comment:
When is `optimizely` a falsy value? Is that when this hook is used and there is no ancestor `OptimizelyProvider`? If that is the reason, the hook will never work as intended I'm guessing. At the very least I think that deserves an error-level log. Is there any other conventional way for a hook to signal that it failed to be setup properly or can't be used in the current context? Maybe return `null`/`undefined`?

Reply:
Discussed IRL that it's not valid for any of our hooks or components to be used outside of an `<OptimizelyProvider>` tree with a valid `optimizely` prop.